### ISSUE TITLE
Message signing with Airgapped HW

### ISSUE THREAD

**Issue by kamordi** [sparrowwallet/sparrow#615]
I tried to sign a message with my SeedSigner, but the button for "Sign" isn't available. I tried to find any hints here in the issues, but wasn't successful. 
I also tried with an USB HW, there the button is available and message signing works. 

As far as I can explain this, the messageSignDialog is called without the active wallet, because the previous check for Keystore only checks for USB_HW. 

Can a support for Airgapped HW  also be added here?

**Comment by craigraw** [comment]
Does SeedSigner support message signing?

**Comment by kamordi** [comment]
Good question, in my naivety I expected it would. But I only was able to find an active issue, asking for this. 

I'm not sure if the proposed flow fits the current flow of Sparrow.

**Comment by craigraw** [comment]
I’m not aware of any QR standards around message signing - would need to be another UR type in the Blockchain Commons registry.

**Comment by thibistaken** [comment]
Is airgapped message signing with a Coldcard HWW supported? I could not find the sign button on the Sparrow GUI with the keystore imported via MicroSD. Thanks!

**Comment by craigraw** [comment]
> Is airgapped message signing with a Coldcard HWW supported?

No, unfortunately not. Coldcard has no support for this.

**Comment by thibistaken** [comment]
Ok, thanks for the heads up. I wasn't aware that Coldcard did not support this feature.

**Comment by kamordi** [comment]
So,  in the past I heard about the development of this kind of feature in the Seedsigner and also in Specter-Wallet. 

https://github.com/SeedSigner/seedsigner/pull/262

https://github.com/cryptoadvance/specter-desktop/pull/1938

This could be a great base for signing with Airgapped HW.

**Comment by newtonick** [comment]
@craigraw we would like add QR message signing to Seedsigner. You can see the format Specter Desktop uses on the PR where this is being worked on (but not in a release yet). https://github.com/SeedSigner/seedsigner/pull/262#issue-1419886558

We'd love to find a way to support QR message signing between Seedsigner and Sparrow. What if we used the format Specter Desktop is using? It assumes a single static frame (not animated) in both directions.

Let me know your thoughts.

**Comment by craigraw** [comment]
@newtonick does it not make more sense to create a UR type for this? It seems reasonable (and in fact common) that a message size will be larger than what can be contained in a single static frame. Whatever we decide on, we'll have to support for a long time, so it may make sense to have a well documented specification that covers a wide set of use cases into the future.

**Comment by newtonick** [comment]
@craigraw a UR type would be great. I’m not going to spend time attempting to document a spec proposal and coordinate/iterate with BC on that proposal. I’m glad folks to this, just not where I want to spend my time. I do agree a common standard would be ideal.

**Comment by RobDoesData** [comment]
> > Is airgapped message signing with a Coldcard HWW supported?
> 
> No, unfortunately not. Coldcard has no support for this.

Cold card does indeed support airgapped message signing, link here: https://coldcard.com/docs/sign-text-file

**Comment by dadofsambonzuki** [comment]
I understand that message signing (using the Specter format) is in the imminent SeedSigner 0.7.0 release.

There are work arounds using external QR generators , but it would be nice to have the flow within Sparrow.

@craigraw - would you consider supporting this static format in the interim (despite its limitations), whilst a dynamic spec is developed between SS, Specter and Sparrow?

**Comment by craigraw** [comment]
Yes, I would consider supporting it, although I do need to look into the Blockchain Commons request/response envelope format again, which would be a more capable approach (albeit more technical to implement). There is an example using this approach here: https://github.com/BlockchainCommons/BCSwiftExampleSigner

**Comment by kdmukai** [comment]
@craigraw @dadofsambonzuki and certainly if and when a more robust spec emerges, SeedSigner already supports different output QR formats for different coordinators so it wouldn't be a problem if Sparrow moves to a format that's different from Specter Desktop.

**Comment by craigraw** [comment]
I've decided to go ahead with the Specter/Seedsigner format, for two reasons:
1. Restrictions in message length are not unusual - for example Coldcard airgapped signing restricts to 240 chars
2. The UR request/response approach is more complex to implement, and thus less likely to see support in the near term

That said, if there is a need to sign longer documents, I'll consider implementing the UR approach, which is a more general approach to calling various functions on a device via QR code.

Implemented in 201900aa - signing by QR is performed using the button on the bottom left of the dialog. The following message was signed using Seedsigner:
 
<img width="656" alt="Screenshot 2023-09-25 at 13 51 00" src="https://github.com/sparrowwallet/sparrow/assets/862166/0492933d-24f5-4704-b2af-3f92a0cb19ca">
